foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/platform/database/PostgreSQL10Platform.java

```diff
@@ -0,0 +1,242 @@
+        if (PgObjectAccessor.PG_OBJECT_ACCESSOR.isInstance(parameter)) {
+            statement.setObject(name, parameter);
+        } else {
+            super.setParameterValueInDatabaseCall(parameter, statement, name, session);
+        }
+    }
+
+    // Postgres specific JSON types support:
+    // Stores JsonValue instances as JSONB.
+    /**
+     * INTERNAL:
+     * Convert JSON value field to JDBC statement type.
+     * Postgres JSON storage type is {@code JSONB} and target Java type is {@code PGobject}.
+     *
+     * @param <T> classification type
+     * @param jsonValue source JSON value field
+     * @return converted JDBC statement type
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T convertJsonValueToDataValue(final JsonValue jsonValue) throws PersistenceException {
+        if (jsonValue == null) {
+            return null;
+        }
+        final String jsonAsString = super.convertJsonValueToDataValue(jsonValue);
+        // Following code is called through reflection to avoid PGobject dependency
+        try {
+            return (T) PgObjectAccessor.PG_OBJECT_ACCESSOR.newPgObject(jsonAsString);
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            throw new PersistenceException("JSON value to database type conversion failed.", e);
+        }
+    }
+
+    // Default resultSet.getString(columnNumber); call works too.
+    /**
+     * Retrieve JSON data from JDBC {@code ResultSet}.
+     * JSON data retrieved from Postgres JDBC {@code ResultSet} are returned as {@code PGobject} instance.
+     * It must be converted to {@code String} first to be accepted by common {@code JsonTypeConverter}.
+     *
+     * @param resultSet source JDBC {@code ResultSet}
+     * @param columnNumber index of column in JDBC {@code ResultSet}
+     * @return JSON data from JDBC {@code ResultSet} as {@code String} to be parsed by common {@code JsonTypeConverter}
+     * @throws SQLException if data could not be retrieved
+     */
+    @Override
+    public Object getJsonDataFromResultSet(final ResultSet resultSet, final int columnNumber) throws SQLException {
+        // ResultSet returns an instance of PGobject.
+        final Object rawData = resultSet.getObject(columnNumber);
+        // Following code is called through reflection to avoid PGobject dependency.
+        if (PgObjectAccessor.PG_OBJECT_ACCESSOR.isInstance(rawData)) {
+            try {
+                return PgObjectAccessor.PG_OBJECT_ACCESSOR.getValue(rawData);
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                throw new PersistenceException("Database PGobject conversion failed.", e);
+            }
+        // Fallback option when String value is returned.
+        } else if (rawData instanceof String) {
+            return rawData;
+        }
+        throw new PersistenceException("Unknown JSON type returned from database");
+    }
+
+}
```

Comment:
PostgreSQL looks like one big workaroud due JDBC drive limits.
Any chance to call DB JSON as a String like MySQL platform?

Reply:
Unfortunately not. At least `public <T> T convertJsonValueToDataValue(final JsonValue jsonValue) throws PersistenceException` with all those reflection calls is a must because `PGobject` must be set into statement.
Theoretically I can  remove `public Object getJsonDataFromResultSet(final ResultSet resultSet, final int columnNumber) throws SQLException` because default code from DatabasePlatform superclass works too. But I'd like to have Statement set and ResultSet get code consistent.